compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -465,6 +465,42 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             );
         }
 
+        if let Mode::MethodCall = mode && let SelfSource::MethodCall(mut rcvr_expr) = source {
+        while let hir::ExprKind::MethodCall(_path_segment, parent_expr, _args, span) =
+        rcvr_expr.kind
+        {
+            rcvr_expr = parent_expr;
+            let prev_ty = self.resolve_vars_if_possible(
+                self.typeck_results
+                    .borrow()
+                    .expr_ty_adjusted_opt(rcvr_expr)
+                    .unwrap_or(Ty::new_misc_error(self.tcx)),
+            );
+            /* err.span_note(
+                span,
+                format!("TEST: the method `{item_name}` is implemented on `{rcvr_expr}`"),
+            ); */
+
+            if let Ok(_pick) = self.probe_for_name(
```

Comment:
Why is this `probe_for_name` not picking up B?

Reply:
what's the issue you're seeing exactly? What do you mean by `not picking up B`?

Reply:
I am expecting the `probe_for_name` to find `struct B` as the receiver for the function `foo`. I am using the while loop to walk up the inheritance path to find the method `foo` in any of the predecessors. Is that making things clear?

Reply:
For debugging, you can do `err.span_node(span, format!("{prev_ty:?}"));` to unconditionally show what type each method call segment is found to be, in order to confirm that the logic before `probe_for_name` is correct. This all looks correct to me at first glance, but I wonder if there might be some subtle bug that we are missing.

Reply:
The thing I can think of is that maybe `expr_ty_adjusted_opt` is not finding a type for the `rcvr_expr`, but that's speculation.